src/Adapters/Auth/oauth2.js

```diff
@@ -0,0 +1,155 @@
+function validateAuthData(authData, options) {
+  return requestTokenInfo(options, authData.access_token).then(response => {
+    if (
+      response &&
+      response.active &&
+      (!options ||
+        !options.hasOwnProperty('useridField') ||
+        !options.useridField ||
+        authData.id == response[options.useridField])
+    ) {
+      return;
+    }
+    const errorMessage = 'OAuth2 access token is invalid for this user.';
+    logger.error(errorMessage);
+    throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, errorMessage);
+  });
+}
+
+function validateAppId(appIds, authData, options) {
+  if (
+    !(options && options.hasOwnProperty('appidField') && options.appidField)
+  ) {
+    return Promise.resolve();
+  } else {
+    if (!appIds.length) {
+      const errorMessage =
+        'OAuth2 configuration is missing the client app IDs ("appIds" config parameter).';
+      logger.error(errorMessage);
+      throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, errorMessage);
+    }
+    return requestTokenInfo(options, authData.access_token).then(response => {
+      const appidField = options.appidField;
+      if (response && response[appidField]) {
+        const responseValue = response[appidField];
+        if (Array.isArray(responseValue)) {
+          if (
+            typeof responseValue.find(function(element) {
+              return appIds.includes(element);
+            }) !== 'undefined'
+          ) {
+            return;
+          }
+        } else {
+          if (appIds.includes(responseValue)) {
+            return;
+          }
+        }
+      }
+      const errorMessage2 =
+        "OAuth2: the access_token's appID is empty or is not in the list of permitted appIDs in the auth configuration.";
+      logger.error(errorMessage2);
+      throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, errorMessage2);
+    });
+  }
+}
+
+// A promise wrapper for requests to the OAuth2 token introspection endpoint.
+function requestTokenInfo(options, access_token) {
+  return new Promise(() => {
+    if (!options || !options.tokenIntrospectionEndpointUrl) {
+      const errorMessage =
+        'OAuth2 token introspection endpoint URL is missing from configuration!';
+      logger.error(errorMessage);
+      throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, errorMessage);
+    }
+    const parsedUrl = url.parse(options.tokenIntrospectionEndpointUrl);
+    const postData = querystring.stringify({
+      token: access_token,
+    });
+    const headers = {
+      'Content-Type': 'application/x-www-form-urlencoded',
+      'Content-Length': Buffer.byteLength(postData),
+    };
+    if (options.authorizationHeader) {
+      headers['Authorization'] = options.authorizationHeader;
+    }
+    const postOptions = {
+      hostname: parsedUrl.hostname,
+      path: parsedUrl.pathname,
+      method: 'POST',
+      headers: headers,
+    };
+    return httpsRequest.request(postOptions, postData);
```

Comment:
You can mock httpsRequest to test the results from `requestTokenInfo` similar to 

https://github.com/parse-community/parse-server/blob/d83a0b680854f52ce30205d3558c0133d7f8cf19/spec/AuthenticationAdapters.spec.js#L527